### PR TITLE
Test to check JMX metrics

### PR DESCRIPTION
**Pull request by steveandwang**
Test to see whether JMX changes as you'd expect it to after running C*.

### CODE REVIEW

File: jmxmetrics_test.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                before.append(jmx.read_attribute(mbean, attribute))
+
+            node.stress(['write', 'n=100K', '-port jmx=7100'])
+
+            attr_counter = 0
+            for package, bean, attribute, expected in MBEAN_VALUES:
+                a_value = jmx.read_attribute(mbeans[attr_counter], attribute)
+                b_value = before[attr_counter]
+                if expected == 'MBeanIncrement':
+                    if b_value > a_value:
```
Comment (EnigmaCurry) [inline jmxmetrics_test.py]: I think we want >= here

File: jmxmetrics_test.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                        errors += mbeans[attr_counter] + " has a before value of " + str(b_value) + " and after value of " + str(a_value) + " and did not increment" + "\n"
```
Comment (EnigmaCurry) [inline jmxmetrics_test.py]: String concatentation is usually an expensive operation. Better to make errors a list and then append the errors to that. If you still want it to be a string at the end, you can do : 

    errors = "\n".join(errors)

File: jmxmetrics_test.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                elif expected == 'MBeanZero':
+                    if not (b_value == 0 and a_value == 0):
+                        errors += mbeans[attr_counter] + " has a before value of " + str(b_value) + " and after value of " + str(a_value) + " and they do not equal zero" + "\n"
+                # If expected is none of the above, then expected should be a number.
+                else:
+                    if a_value != expected:
+                        errors += mbeans[attr_counter] + " has an after value of " + str(a_value) + " which does not equal " + str(expected) + "\n"
+                attr_counter += 1
+
+            assert len(errors) == 0, errors
```
Comment (EnigmaCurry) [inline jmxmetrics_test.py]: assert works, but it doesn't print as nice a message as the unittest assert methods do. Better to do:

    self.assertEqual(len(errors), 0)

**Comment by EnigmaCurry** [comment]
I haven't run this yet, but it looks good.

**Comment by EnigmaCurry** [comment]
Find some other metrics for decrement, equal, and zero so we're sure the other tests work too, then looks good to merge to me.

File: jmxmetrics_test.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                ('metrics', ('ColumnFamily', 'LiveSSTableCount'), 'Value', 'MBeanIncrement'), 
+                ('metrics', ('ColumnFamily', 'MemtableColumnsCount'), 'Value', 'MBeanIncrement'),
+                ('metrics', ('ColumnFamily', 'MemtableLiveDataSize'), 'Value', 'MBeanIncrement'),
+                ('metrics', ('ColumnFamily', 'MemtableOnHeapSize'), 'Value', 'MBeanIncrement'),
+                ('metrics', ('ColumnFamily', 'MemtableSwitchCount'), 'Value', 'MBeanIncrement'),
+                ('metrics', ('ColumnFamily', 'PendingFlushes'), 'Value', 'MBeanIncrement'),
+                ('db', 'IndexSummaries', 'MemoryPoolSizeInMB', 'MBeanIncrement'),
+                ('db', 'IndexSummaries', 'IndexIntervals', 'MBeanIncrement')]
+
+class JMXValueTest(Tester):
```
Comment (ptnapoleon) [inline jmxmetrics_test.py]: For consistency's sake with the other files, this should be `TestJMXMetrics`

File: jmxmetrics_test.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+    def __init__(self, *args, **kwargs):
+        Tester.__init__(self, *args, **kwargs)
+
+    def begin_test(self):
```
Comment (ptnapoleon) [inline jmxmetrics_test.py]: Add some doxygen markup on this, please and thanks. Include at least the `@jira_ticket` tag. Also, could you give the test a more descriptive name?

File: jmxmetrics_test.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            for package, bean, attribute, expected in MBEAN_VALUES:
+                # In the case that the file name is longer, then we put it in the form of tuple in bean.
+                if type(bean) == tuple:
+                    mbean = make_mbean(package, type = bean[0], name = bean[1])
+                else:
+                    mbean = make_mbean(package, bean)
+                mbeans.append(mbean)
+                before.append(jmx.read_attribute(mbean, attribute))
+
+            node.stress(['write', 'n=100K', '-port jmx=7100'])
```
Comment (ptnapoleon) [inline jmxmetrics_test.py]: Is this test going to run on 2.0 as well? If so, this stress call will fail.

File: jmxmetrics_test.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+class JMXValueTest(Tester):
+
+    def __init__(self, *args, **kwargs):
+        Tester.__init__(self, *args, **kwargs)
+
+    def begin_test(self):
+        cluster = self.cluster
+        cluster.populate(1)
+        node = cluster.nodelist()[0]
+        remove_perf_disable_shared_mem(node)
```
Comment (ptnapoleon) [inline jmxmetrics_test.py]: Could you leave a comment explaining why you did this? It's obvious to anyone familiar with the issue, but not everyone reading the test will be.

File: jmxmetrics_test.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                ('metrics', ('ColumnFamily', 'MemtableLiveDataSize'), 'Value', 'MBeanIncrement'),
+                ('metrics', ('ColumnFamily', 'MemtableOnHeapSize'), 'Value', 'MBeanIncrement'),
+                ('metrics', ('ColumnFamily', 'MemtableSwitchCount'), 'Value', 'MBeanIncrement'),
+                ('metrics', ('ColumnFamily', 'PendingFlushes'), 'Value', 'MBeanIncrement'),
+                ('db', 'IndexSummaries', 'MemoryPoolSizeInMB', 'MBeanIncrement'),
+                ('db', 'IndexSummaries', 'IndexIntervals', 'MBeanIncrement')]
+
+class JMXValueTest(Tester):
+
+    def __init__(self, *args, **kwargs):
```
Comment (ptnapoleon) [inline jmxmetrics_test.py]: This is totally redundant, no need to include this `__init__` if you aren't modifying anything.

**Comment by stef1927** [comment]
Please check [CASSANDRA-9448] (https://issues.apache.org/jira/browse/CASSANDRA-9448), some of the metrics names will change in 3.0.

File: jmxmetrics_test.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+# MBEAN_VALUES_POST_3.0 = 
+
+class TestJMXMetrics(Tester):
+
+    def __init__(self, *args, **kwargs):
+        Tester.__init__(self, *args, **kwargs)
+
+    def begin_test(self):
+        """
+        @jira_ticket 7436
```
Comment (ptnapoleon) [inline jmxmetrics_test.py]: Should be `@jira_ticket CASSANDRA-7436` please

File: jmxmetrics_test.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+                # In the case that the file name is longer, then we put it in the form of tuple in bean.
+                if type(bean) == tuple:
+                    mbean = make_mbean(package, type = bean[0], name = bean[1])
+                else:
+                    mbean = make_mbean(package, bean)
+                mbeans.append(mbean)
+                before.append(jmx.read_attribute(mbean, attribute))
+
+            if cluster.version() < "2.1":
+                node.stress(['-o', 'insert', '-n', '100', '-p', '7100'])
```
Comment (ptnapoleon) [inline jmxmetrics_test.py]: You're inserting 100K lines in 2.1, but only 100 in 2.0

**Comment by ptnapoleon** [comment]
+1! Great job! :+1: :fireworks:

**Pull request merged**